upnpdiscovery/service.go

```diff
@@ -0,0 +1,229 @@
+		}
+	}
+	return queries, endpointMap
+}
+
+// DiscoverResources discovers different rtsp cameras that use onvif.
+func (dis *upnpDiscovery) DiscoverResources(ctx context.Context, extra map[string]any) ([]resource.Config, error) {
+	cams := []resource.Config{}
+
+	discoverQueries := dis.queries
+
+	extraQuery, ok := getQueryFromExtra(extra)
+	if ok {
+		discoverQueries = append(discoverQueries, extraQuery)
+	}
+	// worth noting that the discovered hosts are not guaranteed to be rtsp camera.
+	// We assume that the user knew what they were looking for.
+	hosts, hostmap, err := viamupnp.FindHost(ctx, dis.logger, discoverQueries, dis.rootOnly)
+	if err != nil {
+		return nil, err
+	}
+
+	for hostNum, host := range hosts {
+		query := hostmap[host]
+		endpoints, ok := dis.endpointMap[query]
+		// the query had no endpoints
+		if !ok {
+			camConfig, err := createCameraConfig(createCameraName(hostNum, -1, query), "rtsp://"+host, query)
+			if err != nil {
+				return nil, err
+			}
+			cams = append(cams, camConfig)
+			continue
+		}
+		for endpointNum, endpoint := range endpoints {
+			camConfig, err := createCameraConfig(createCameraName(hostNum, endpointNum, query),
+				fmt.Sprintf("rtsp://%s:%s", host, endpoint), query)
+			if err != nil {
+				return nil, err
+			}
+			cams = append(cams, camConfig)
+		}
+	}
+
+	return cams, nil
+}
+
+// regex to remove non alpha numerics.
+var reg = regexp.MustCompile("[^a-zA-Z0-9]+")
+
+// createCameraName creates a camera name based on the query.
+func createCameraName(hostNum, endpointNum int, query viamupnp.DeviceQuery) string {
+	camName := fmt.Sprintf("camera%v", hostNum)
+	if endpointNum != -1 {
+		camName = fmt.Sprintf("%s-endpoint%v", camName, endpointNum)
+	}
+	if stripManufacturer := reg.ReplaceAllString(query.Manufacturer, ""); stripManufacturer != "" {
+		camName = fmt.Sprintf("%s-%s", camName, stripManufacturer)
+	}
+	if stripModel := reg.ReplaceAllString(query.ModelName, ""); stripModel != "" {
+		camName = fmt.Sprintf("%s-%s", camName, stripModel)
+	}
+	if stripSerial := reg.ReplaceAllString(query.SerialNumber, ""); stripSerial != "" {
+		camName = fmt.Sprintf("%s-%s", camName, stripSerial)
+	}
+	return camName
+}
+
+func createCameraConfig(name, address string, query viamupnp.DeviceQuery) (resource.Config, error) {
+	// using the camera's Config struct in case a breaking change occurs
+	_true := true
+	attributes := viamrtsp.Config{Address: address, Query: query, RTPPassthrough: &_true}
+	var result map[string]interface{}
+
+	// marshal to bytes
+	jsonBytes, err := json.Marshal(attributes)
+	if err != nil {
+		return resource.Config{}, err
+	}
+
+	// convert to map to be used as attributes in resource.Config
+	if err = json.Unmarshal(jsonBytes, &result); err != nil {
+		return resource.Config{}, err
+	}
+
+	return resource.Config{
+		Name: name, API: camera.API, Model: viamrtsp.ModelAgnostic,
+		Attributes: result, ConvertedAttributes: &attributes,
+	}, nil
+}
```

Comment:
can make this common to discovery services - we should move it to the main rtsp package.

Reply:
the behavior is different on this technically since I wanted to return the query in the attributes as well to inform users how the device was found.

Can remove the query and make this one function if requested.